xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -1302,10 +1301,17 @@ public void run() {
     }
   }
 
-  private final class AdsStream implements StreamObserver<DiscoveryResponse> {
+  private enum ResourceType {
+    LDS, RDS, CDS, EDS
+  }
+
+  private final class AdsStream {
```

Comment:
This is a mess. You should implement separate AdsStreams, which are for separate `StreamObserver<T>`. One for v2 and another for v3, instead of putting everything together and use one stub or another. 

Then XdsClientImpl should have a bunch of top-level xDS response handlers with nice and clean method overloading. Something like this:

```
private void handleLdsResponse(...v2... response);

private void handleLdsResponse(...v3... response);

private void handleRdsResponse(...v2... response);

private void handleRdsResponse(...v3... response);

// ... same for CDS and EDS

```
This main body of these handlers will just convert proto messages (v2 and v3) to general types (Java objects) defined in `EnvoyProtoData` and the whole business logic will be done in helper methods that take in converted data objects only.

This would be a significant work to keep things well organized, instead of dump another copy of Stub/StreamObserver, etc. The easier step we take for now, the harder situation we will be in the future maintenance/debug work.

I'd suggest first come up with a draft with components/methods reorganizations.

Reply:
> You should implement separate AdsStreams, which are for separate StreamObserver<T>. One for v2 and another for v3, instead of putting everything together and use one stub or another.

I don't see essential difference

 having`AdsStream.stubV2` and `AdsStream.stubV3` 

vs 

having `XdsClientImpl.adsStreamV2` and `XdsClientImpl.adsStreamV3` instead of putting everything together and use one stub or another,

or even vs 

`XdsClientImplV2` and `XdsClientImplV3` instead of putting everything together and using one AdsStream or another.

Can you explain more? 

The downside of having `XdsClientImpl.adsStreamV2` and `XdsClientImpl.adsStreamV3` is: you have a lot of duplicate identical fields and private methods for AdsStreamV2 and AdsStreamV3. Maybe introducing an abstract AdsStream/abstract XdsClientImpl (If we could know the version ahead of time) to share the common code could help.

Reply:
> Something like this:
>```java
>private void handleLdsResponse(...v2... response);
>private void handleLdsResponse(...v3... response);
>private void handleRdsResponse(...v2... response);
>private void handleRdsResponse(...v3... response);
>// ... same for CDS and EDS
>```

Probably we want the outer class `XdsClientIml` be version/proto agnostic, something like this:
```java
private void handleLdsResponse(EnvoyProtoData.DiscoveryResponse response);
```

Reply:
> I don't see essential difference
> 
> having`AdsStream.stubV2` and `AdsStream.stubV3`
> 
> vs
> 
> having `XdsClientImpl.adsStreamV2` and `XdsClientImpl.adsStreamV3` instead of putting everything together and use one stub or another,
> 
> or even vs
> 
> `XdsClientImplV2` and `XdsClientImplV3` instead of putting everything together and using one AdsStream or another.


You can define an abstract class and implement two sub variants as the AdsStream. Then by any chance, XdsClientImpl maintains and create only one of them, but not having two fields while one being null. Need to come with an elegant design instead of dumping every thing together.

Reply:
Refactored. It's another big refactor :(. Note that support for two versions may be deleted in the future. Does it really worth the effort?

Reply:
> Refactored. It's another big refactor :(. Note that support for two versions may be deleted in the future. Does it really worth the effort?

I thought support for 2 versions is going to be there for the foreseeable future as per the versioning doc unless I missed some refs. Can you cite a reference?

Reply:
IMO, by any means we should care about code qualities, expecting things to be changes should not be a reason to leave messy code/design in the production codebase. More changes may be made based on code being checked, even if it contains temporary parts. Deleting the temporary things in the future would take another round of refactoring. It's never going to be a simple undo/revert. So leaving such a big change without considering the overall code quality makes future refactors even harder.